bpf/lib/nodeport.h

```diff
@@ -832,6 +833,81 @@ int tail_nodeport_nat_egress_ipv6(struct __ctx_buff *ctx)
 	return send_drop_notify_error_ext(ctx, 0, ret, ext_err, CTX_ACT_DROP, METRIC_EGRESS);
 }
 
+declare_tailcall_if(__not(is_defined(IS_BPF_LXC)), CILIUM_CALL_IPV6_NODEPORT_DSR_INGRESS)
+int tail_nodeport_dsr_ingress_ipv6(struct __ctx_buff *ctx)
+{
+	struct ct_state ct_state_new = {};
+	struct ipv6_ct_tuple tuple = {};
+	struct ct_state ct_state = {};
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	__u32 monitor = 0;
+	union v6addr addr;
+	int ret, l4_off;
+	__be16 port;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip6)) {
+		ret = DROP_INVALID;
+		goto drop_err;
+	}
+
+	ret = lb6_extract_tuple(ctx, ip6, &l4_off, &tuple);
+	if (IS_ERR(ret))
+		goto drop_err;
+
+	addr.p1 = ctx_load_meta(ctx, CB_ADDR_V6_1);
+	addr.p2 = ctx_load_meta(ctx, CB_ADDR_V6_2);
+	addr.p3 = ctx_load_meta(ctx, CB_ADDR_V6_3);
+	addr.p4 = ctx_load_meta(ctx, CB_ADDR_V6_4);
+	port = (__be16)ctx_load_meta(ctx, CB_PORT);
+
+	ctx_store_meta(ctx, CB_PORT, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_1, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_2, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_3, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_4, 0);
+
+	ret = ct_lb_lookup6(get_ct_map6(&tuple), &tuple, ctx, l4_off,
+			    CT_EGRESS, &ct_state, &monitor);
+	switch (ret) {
+	case CT_NEW:
+	case CT_REOPENED:
+create_ct:
+		ct_state_new.src_sec_id = WORLD_ID;
+		ct_state_new.dsr = 1;
+		ct_state_new.ifindex = (__u16)NATIVE_DEV_IFINDEX;
```

Comment:
AFAIU, this code comes to replace the code you've removed from bpf_lxc, right? In bpf_lxc, we didn't set `ct_state_new.ifindex` AFAICS. Why do we need to set it here?

Reply:
Right now it's essentially prep-work, to keep the CT entry on par with that we build for a local backend.

My gut feeling is that with multiple external interfaces, we'll end up needing an additional FIB-lookup **after** the RevDNAT in `to-netdev` happened. So let's make sure that we already create CT entries that have all the needed information for that case. Alternatively we clean it up consistently for both DSR and local-backend sometime in the future.

Reply:
Got it! Not necessary today, but added for consistency :+1: